### DISCUSSION TITLE
Too many files loaded by next.js and some of them are so thin

### DISCUSSION THREAD

**Discussion by pooriamo** [vercel/next.js#11082] [Help]
These are the (huge number of) files that are loaded in my home page:

![image](https://user-images.githubusercontent.com/9452720/76699764-e84a5f00-66c5-11ea-999e-eeae88669fd3.png)

is this really normal?! 

Also the last to files contain only one small line, loading them separately makes the performance really poor.

**Comment by timneutkens** [comment]
You can read more about it here: https://nextjs.org/blog/next-9-2#improved-code-splitting-strategy

**Reply by pooriamo** [reply]
@timneutkens  Unfortunately nextjs.org is on a server that has blocked my country and my vpn is not working for now, I'll look at it later and If I had any questions I'll ask again. Thanks for your reply Tim.

**Reply by janhoogeveen** [reply]
> @timneutkens Unfortunately nextjs.org is on a server that has blocked my country and my vpn is not working for now, I'll look at it later and If I had any questions I'll ask again. Thanks for your reply Tim.

Improved Code-Splitting Strategy
Next.js versions prior to 9.2 had a fixed set of JavaScript bundles required to load and make a page interactive:

The page's JavaScript file
A file with common JavaScript
Next.js client-side runtime bundle
Webpack client-side runtime bundle
Dynamic imports (added through next/dynamic, when used)
To make the page interactive, all of these bundles have to load as they depend on each other to boot up React in the browser.

Because all of these bundles are required for the application to become interactive, it's important they are as optimized as possible. In practice, this means not over-downloading code from other parts of the application.

For this reason, Next.js used a commons bundle that held common JavaScript between pages. The calculation of the old bundle-splitting strategy to generate commons was a usage-ratio heuristic. If a module was used in more than 50% of all pages, it'd be marked as a common module. Otherwise, it was bundled in the page's JavaScript file.

However, applications can consist of many different types of pages. For example, marketing pages, a blog, and a dashboard. If there was a large number of marketing pages compared to other page types, the commons calculation would result in optimizations heavily focused on the marketing pages.

Our goal is to optimize for all page types, in one application.

Alex Castle proposed a new method of chunking (creation of separate JavaScript files) that allows optimized commons chunking with multiple files, including when many page types are involved.

Today, we're happy to announce this new chunking behavior is enabled by default in Next.js 9.2. We'd like to extend deep gratitude to the Google Chrome team and Alex Castle for contributing this change. This change reflects the cumulative effort of weeks of research, lab testing, real-world testing, and implementation.

The new chunking implementation leverages HTTP/2 to deliver a greater number of smaller sized chunks.

Under the new heuristic, chunks are created for:

A minimal chunk for each page.
A framework chunk containing React, ReactDOM, React's Scheduler, etc.
Library chunks for any node_module dependency over 160kb (pre-minify/gzip)
A commons chunk for code used across all pages.
As many shared chunks (used by 2 or more pages) as possible, optimizing for overall application size and initial load speed.
Next.js' client-side runtime.
Webpack runtime.

**Reply by pooriamo** [reply]
@janhoogeveen You're my hero <3

**Reply by pooriamo** [reply]
@timneutkens Thanks Tim. I read the article. But I still don't get it. Isn't 23 files counted as "A lot"?
Specially the last two files which only contain a few words. Couldn't we inline them instead of doing two separate http requests for just a few words?

What am I missing?

**Reply by Timer** [reply]
HTTP/2 performs best at 25 or less files, so this is within the threshold.

Also, those last two files are not required for your application to hydrate so they don't cause any performance drop!
If you look at the preload tags, you'll see that the last two (SSG and build manifest) are not preloaded and thus handled last/with low priority.

**Reply by pooriamo** [reply]
@Timer Thanks Joe. I get it now. It's an honor for me to talk to you guys.

**Reply by FBosler** [reply]
@timneutkens I am observing strange behavior in my app and am not sure where exactly this is coming from. Say for example [this page](https://www.viteach.de/node/5f437c7192193f7caf74c24d)

Has some pretty huge bundles:
![image](https://user-images.githubusercontent.com/14016698/106623111-8fe9ad80-6574-11eb-9e60-a1b41da21dc0.png)
as created by `npx bundle-wizard https://www.viteach.de/node/5f437c7192193f7caf74c24d`

On the page itself, it seems that only a fraction of the js in the commons.js is actually being used:
![image](https://user-images.githubusercontent.com/14016698/106623404-dccd8400-6574-11eb-98b1-144ea4f8b236.png)

I would expect the commons chunk to only include the code that is actually used across all pages (i.e. coverage should be pretty high on every page or am I missing something?)

Also, any idea what the culprit could be for the code to be used across all pages? The vast majority of all the mentioned libraries are not used on `_app` or `_document.js` which I understand would be all libraries.

One thing that comes to mind is my large `dynamicComponent`-Map:

```react.js
import dynamic from 'next/dynamic'

const COMPONENT_MAP = {
    '5ed36c8b7b6a7f31610eb0a9': {
        path: './Fractions/BaseDefinitions',
        module: dynamic(import('./Fractions/BaseDefinitions')),
    },
    ...,
    '5ed3719553181884ce6f454f': {
        path: './Fractions/AdditionSubtraction/AdditionSameDenom',
        module: dynamic(import('./Fractions/AdditionSubtraction/AdditionSameDenom')),
    },
}

const dynamicComponent = (nodeId, props) => {
    const { module: Component } = COMPONENT_MAP[nodeId] || {}

    if (Component) {
        return <Component {...{ ...props, nodeId }} />
    } else {
        return <div>{`Could not find component for ${nodeId} please check nodes`}</div>
    }
}

export { dynamicComponent }
```

Said page is built dynamically, i.e. I create a path for every id via `getStaticPaths` and get some additional SEO information in `getStaticProps` and then render them out via a lookup through `dynamicComponent`. Since my dynamicComponents includes every component and therefore every library, could this be the reason? If so, how would I go about changing it?

**Comment by sahanatvessel** [comment]
@leerob this might still be happening on Vercel + next?
